Cannot run as root user with non blank password

I have noticed that it is not possible to connect to embedded MySQL using the username root and providing a non empty string password. It works fine if the password is empty string but if I use anything other value it fails to connect.

Is this because MySQL does not allow you to have a root user with a blank password? I suspect that it is, in which case feel free to close this issue but I just wanted to check whether it was something you knew about already?

I have created a repo here which demonstrates the issue, there are 4 tests, 3 work fine and 1 fails.

https://github.com/michaelruocco/embedded-mysql-password-issue

When the test fails it throws the following error when trying to connect to the embedded MySQL instance.

```
java.sql.SQLException: Cannot create PoolableConnectionFactory (Access denied for user 'root'@'localhost' (using password: YES))
	at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:2294)
	at org.apache.commons.dbcp2.BasicDataSource.createDataSource(BasicDataSource.java:2039)
	at org.apache.commons.dbcp2.BasicDataSource.getConnection(BasicDataSource.java:1533)
	at uk.co.tpplc.PasswordTest.testConnection(PasswordTest.java:60)
	at uk.co.tpplc.PasswordTest.shouldConnectWithRootUserAndPassword(PasswordTest.java:35)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:160)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:117)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:42)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:253)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:84)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
Caused by: java.sql.SQLException: Access denied for user 'root'@'localhost' (using password: YES)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:695)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:663)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:653)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:115)
	at com.mysql.cj.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:1663)
	at com.mysql.cj.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:662)
	at com.mysql.cj.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:352)
	at com.mysql.cj.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:221)
	at org.apache.commons.dbcp2.DriverConnectionFactory.createConnection(DriverConnectionFactory.java:39)
	at org.apache.commons.dbcp2.PoolableConnectionFactory.makeObject(PoolableConnectionFactory.java:256)
	at org.apache.commons.dbcp2.BasicDataSource.validateConnectionFactory(BasicDataSource.java:2304)
	at org.apache.commons.dbcp2.BasicDataSource.createPoolableConnectionFactory(BasicDataSource.java:2290)
	... 31 more
```

--- a/src/main/java/com/wix/mysql/config/MysqldConfig.java
+++ b/src/main/java/com/wix/mysql/config/MysqldConfig.java
@@ -34,6 +34,10 @@ public class MysqldConfig extends ExecutableProcessConfig {
             }
         });
 
+        if (user.name.equals("root")) {
+            throw new IllegalArgumentException("Usage of username 'root' is forbidden as it's reserved for system use");
+        }
+
         this.port = port;
         this.charset = charset;
         this.user = user;

--- a/src/test/scala/com/wix/mysql/MysqldConfigTest.scala
+++ b/src/test/scala/com/wix/mysql/MysqldConfigTest.scala
@@ -37,7 +37,7 @@ class MysqldConfigTest extends SpecWithJUnit {
       mysqldConfig.getTimeZone mustEqual TimeZone.getTimeZone("Europe/Vilnius")
     }
 
-    "fail if building with user 'Root'" in {
+    "fail if building with user 'root'" in {
       aMysqldConfig(v5_6_latest)
         .withUser("root", "doesnotmatter")
         .build() must throwA[IllegalArgumentException](message = "Usage of username 'root' is forbidden")

--- a/src/test/scala/com/wix/mysql/MysqldConfigTest.scala
+++ b/src/test/scala/com/wix/mysql/MysqldConfigTest.scala
@@ -36,5 +36,13 @@ class MysqldConfigTest extends SpecWithJUnit {
       mysqldConfig.getPassword mustEqual "otherpassword"
       mysqldConfig.getTimeZone mustEqual TimeZone.getTimeZone("Europe/Vilnius")
     }
+
+    "fail if building with user 'Root'" in {
+      aMysqldConfig(v5_6_latest)
+        .withUser("root", "doesnotmatter")
+        .build() must throwA[IllegalArgumentException](message = "Usage of username 'root' is forbidden")
+
+    }
+
   }
 }

